backend/src/plonk/prover.rs

```diff
@@ -0,0 +1,743 @@
+        R: RngCore,
+        T: TranscriptWrite<Scheme::Curve, E>,
+    > ProverV2Single<'a, 'params, Scheme, P, E, R, T>
+{
+    /// Create a new prover object
+    pub fn new(
+        params: &'params Scheme::ParamsProver,
+        pk: &'a ProvingKey<Scheme::Curve>,
+        // TODO: If this was a vector the usage would be simpler
+        instance: &[&[Scheme::Scalar]],
+        rng: R,
+        transcript: &'a mut T,
+    ) -> Result<Self, Error>
+    // TODO: Can I move this `where` to the struct definition?
+    where
+        Scheme::Scalar: WithSmallOrderMulGroup<3> + FromUniformBytes<64>,
+    {
+        Ok(Self(ProverV2::new(
+            params,
+            pk,
+            &[instance],
+            rng,
+            transcript,
+        )?))
+    }
+
+    /// Commit the `witness` at `phase` and return the challenges after `phase`.
+    pub fn commit_phase(
+        &mut self,
+        phase: u8,
+        witness: Vec<Option<Vec<Assigned<Scheme::Scalar>>>>,
+    ) -> Result<HashMap<usize, Scheme::Scalar>, Error>
+    where
+        Scheme::Scalar: WithSmallOrderMulGroup<3> + FromUniformBytes<64>,
+    {
+        self.0.commit_phase(phase, vec![witness])
+    }
+
+    /// Finalizes the proof creation.
+    pub fn create_proof(self) -> Result<(), Error>
+    where
+        Scheme::Scalar: WithSmallOrderMulGroup<3> + FromUniformBytes<64>,
+    {
+        self.0.create_proof()
+    }
+}
+
+/// The prover object used to create proofs interactively by passing the witnesses to commit at
+/// each phase.  This supports batch proving.
+#[derive(Debug)]
+pub struct ProverV2<
+    'a,
+    'params,
+    Scheme: CommitmentScheme,
+    P: Prover<'params, Scheme>,
+    E: EncodedChallenge<Scheme::Curve>,
+    R: RngCore,
+    T: TranscriptWrite<Scheme::Curve, E>,
+> {
+    // Circuit and setup fields
+    params: &'params Scheme::ParamsProver,
+    pk: &'a ProvingKey<Scheme::Curve>,
+    // TODO: Add getter
+    pub phases: Vec<sealed::Phase>,
+    // State
+    instance: Vec<InstanceSingle<Scheme::Curve>>,
+    advice: Vec<AdviceSingle<Scheme::Curve, LagrangeCoeff>>,
+    challenges: HashMap<usize, Scheme::Scalar>,
+    next_phase_index: usize,
+    rng: R,
+    transcript: &'a mut T,
+    _marker: std::marker::PhantomData<(P, E)>,
+}
+
+impl<
+        'a,
+        'params,
+        Scheme: CommitmentScheme,
+        P: Prover<'params, Scheme>,
+        E: EncodedChallenge<Scheme::Curve>,
+        R: RngCore,
+        T: TranscriptWrite<Scheme::Curve, E>,
+    > ProverV2<'a, 'params, Scheme, P, E, R, T>
+{
+    /// Create a new prover object
+    pub fn new(
+        params: &'params Scheme::ParamsProver,
+        pk: &'a ProvingKey<Scheme::Curve>,
+        // TODO: If this was a vector the usage would be simpler
+        instances: &[&[&[Scheme::Scalar]]],
+        rng: R,
+        transcript: &'a mut T,
+    ) -> Result<Self, Error>
+    // TODO: Can I move this `where` to the struct definition?
+    where
```

Comment:
I will remove the comment as I didn't find a way to do it.

Reply:
I think we can just remove the comment without issue, because it's a tiny thing.
Addressed in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/acb13f81ff90b81eae1babff68065f814c79c163